presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java

```diff
@@ -138,186 +75,42 @@ public OrcRecordReader(
+                // ListStreamReader and StructStreamReader) instance sizes were not counted, because calling setBytes() in
+                // their constructors is confusing.
+                createStreamReaders(orcDataSource, types, hiveStorageTimeZone, includedColumns, systemMemoryUsage.newAggregatedMemoryContext()),
+                predicate,
+                numberOfRows,
+                fileStripes,
+                fileStats,
+                stripeStats,
                 orcDataSource,
-                decompressor,
+                splitOffset,
+                splitLength,
                 types,
-                this.presentColumns,
+                decompressor,
                 rowsInRowGroup,
-                predicate,
+                hiveStorageTimeZone,
                 hiveWriterVersion,
                 metadataReader,
-                writeValidation);
-
-        streamReaders = createStreamReaders(orcDataSource, types, hiveStorageTimeZone, presentColumnsAndTypes.build(), streamReadersSystemMemoryContext);
-        maxBytesPerCell = new long[streamReaders.length];
-        nextBatchSize = initialBatchSize;
-    }
-
-    private static boolean splitContainsStripe(long splitOffset, long splitLength, StripeInformation stripe)
-    {
-        long splitEndOffset = splitOffset + splitLength;
-        return splitOffset <= stripe.getOffset() && stripe.getOffset() < splitEndOffset;
-    }
+                maxMergeDistance,
+                tinyStripeThreshold,
+                maxBlockSize,
+                userMetadata,
+                systemMemoryUsage,
+                writeValidation,
+                initialBatchSize);
 
-    private static boolean isStripeIncluded(
-            OrcType rootStructType,
-            StripeInformation stripe,
-            Optional<StripeStatistics> stripeStats,
-            OrcPredicate predicate)
-    {
-        // if there are no stats, include the column
-        if (!stripeStats.isPresent()) {
-            return true;
-        }
-        return predicate.matches(stripe.getNumberOfRows(), getStatisticsByColumnOrdinal(rootStructType, stripeStats.get().getColumnStatistics()));
-    }
+        this.includedColumns = includedColumns;
 
-    @VisibleForTesting
-    static OrcDataSource wrapWithCacheIfTinyStripes(OrcDataSource dataSource, List<StripeInformation> stripes, DataSize maxMergeDistance, DataSize tinyStripeThreshold)
-    {
-        if (dataSource instanceof CachingOrcDataSource) {
-            return dataSource;
-        }
-        for (StripeInformation stripe : stripes) {
-            if (stripe.getTotalLength() > tinyStripeThreshold.toBytes()) {
-                return dataSource;
-            }
-        }
-        return new CachingOrcDataSource(dataSource, createTinyStripesRangeFinder(stripes, maxMergeDistance, tinyStripeThreshold));
-    }
+        batchStreamReaders = Arrays.stream(getStreamReaders())
```

Comment:
Could we parameterize `AbstractOrcRecordReader`?  Like: `abstract class AbstractOrcRecordReader<T extends StreamReader>`.  That way we don't need to cast.